### PR TITLE
Add support for input and gestures for hand controllers

### PR DESCRIPTION
**Pull request by FejZa**
## Overview

Companion PR to https://github.com/XRTK/XRTK-Core/pull/552

### CODE REVIEW

File: XRTK.SDK/Packages/com.xrtk.sdk/Profiles~/Input/DataProviders/HandControllers/Simulated/Left/YawCounterClockwiseProfile.asset.meta [hunk 1]
```diff
@@ -1,5 +1,5 @@
 fileFormatVersion: 2
-guid: b7a023ed9a50a624691ad0b5a440cb3c
+guid: 17f3fbe848f439d42b507e772d4815f2
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Profiles~/Input/DataProviders/HandControllers/Simulated/Left/YawCounterClockwiseProfile.asset.meta:2]: Looks like a lot of guids go regenerated. Any chance we can revert them?

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs [hunk 2]
```diff
@@ -11,11 +11,15 @@ namespace XRTK.SDK.UX.Controllers.Hands
     /// <summary>
     /// Hand controller visualizer visualizing hand joints.
     /// </summary>
-    [System.Runtime.InteropServices.Guid("0B653537-D237-4622-ACC6-E351209A5882")]
+    [System.Runtime.InteropServices.Guid("9335E51F-52AB-48EB-A4DC-25832B6B56D6")]
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs:14]: Why did we change this?

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/BaseHandControllerVisualizer.cs [hunk 3]
```diff
@@ -113,8 +113,8 @@ private void UpdateHandJointTransforms(HandData handData)
                 {
                     var jointTransform = GetOrCreateJointTransform(handJoint);
                     var jointPose = jointPoses[handJoint];
-                    jointTransform.localPosition = jointPose.Position;
-                    jointTransform.localRotation = jointPose.Rotation;
+                    jointTransform.localPosition = jointTransform.parent.InverseTransformPoint(jointPose.Position);
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/BaseHandControllerVisualizer.cs:116]: Is this bc the center position is never updated?

File: XRTK.SDK/Packages/com.xrtk.sdk/Profiles~/Input/DataProviders/HandControllers/Simulated/Left/YawCounterClockwiseProfile.asset.meta (hunk 1 above)
Comment (FejZa) [reply XRTK.SDK/Packages/com.xrtk.sdk/Profiles~/Input/DataProviders/HandControllers/Simulated/Left/YawCounterClockwiseProfile.asset.meta:2]: In this case it's fine. The underlying script has changed and the purpose of the asset as well.

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs (hunk 2 above)
Comment (FejZa) [reply XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs:14]: I for sure didn't change this myself. Not sure what happened here. But we can just revert the GUID.

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/BaseHandControllerVisualizer.cs (hunk 3 above)
Comment (FejZa) [reply XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/BaseHandControllerVisualizer.cs:116]: This is a temporary fix I made to see hands again. I am planning to solve this in a different place, probably converter, so the visualizer always gets the correct poses.

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs [hunk 4]
```diff
@@ -4,17 +4,17 @@
...
-using XRTK.EventDatum.Input;
 
 namespace XRTK.SDK.UX.Controllers.Hands
 {
-    /// <summary>
-    /// Hand controller visualizer visualizing hand joints.
-    /// </summary>
-    [System.Runtime.InteropServices.Guid("0B653537-D237-4622-ACC6-E351209A5882")]
-    public class HandControllerJointsVisualizer : BaseHandControllerVisualizer
+    public class HandControllerJointsVisualizer : MonoBehaviour
```
Comment (StephenHodgson) [inline XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs:10]: We probably still need the guid, no?

Comment (FejZa) [reply XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Controllers/Hands/HandControllerJointsVisualizer.cs:10]: I don't think. This is now just a plain MonoBehaviour component in the SDK. It's not being assigned in the controller visualization profiles.

File: XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Pointers/BaseControllerPointer.cs [hunk 5]
```diff
@@ -184,7 +188,14 @@ protected virtual void OnTriggerEnter(Collider other)
...
                 other.IsValidCollision(PointerRaycastLayerMasksOverride ?? MixedRealityToolkit.InputSystem.FocusProvider.GlobalPointerRaycastLayerMasks))
             {
-                capturedNearInteractionObject = other.gameObject;
+                CapturedNearInteractionObject = other.gameObject;
+
+                // Force update the focus provider so the focused target
+                // gets updated before raising the event. If we don't update
+                // the focus provider here, the event will not be raised on the
+                // capture near interaction object.
+                MixedRealityToolkit.InputSystem.FocusProvider.Update();
```
Comment (FejZa) [inline XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Pointers/BaseControllerPointer.cs:197]: @StephenHodgson when you have some time could you please comment on this. I don't like this change at all but it's the only way to currently make near interaction work.

I didn't see a way to make it work without touching the FocusProviders itself and other components of the input system. In general I think we need to stop making these ad-hoc changes for near interaction and come up with a proper concept and data flow for near interaction to work. It doesn't feel quite there yet. I think what you did so far is a good starting point but upon debugging I found there is still more work to do to make this work. It's not enough to just register `OnTriggerEnter()` etc. We need to properly handle the focus change as well.

Comment (FejZa) [reply XRTK.SDK/Packages/com.xrtk.sdk/Runtime/Features/UX/Scripts/Pointers/BaseControllerPointer.cs:197]: Essentially what happens here is that I have to force an update / raycast on the focus provider before `RaiseOnInput...`. If I don't do this, the input event will not be raised on `CapturedNearInteractionObject` but instead on whatever the pointer was focusing before.

**Review by SimonDarksideJ — APPROVED** [review]

**Pull request merged**